Enforce minimum resource granularity

timstclair opened the pull request:
Enforce minimum resource granularity for Storage, Memory, and CPU
resource types. For Storage and Memory, validate that the provided value
is an integer quantity. For CPU, round up quantity to milli-core precesion.

~~Follow up to https://github.com/kubernetes/kubernetes/pull/17548 (which must be merged first).~~

cc/ @thockin @bgrant0607

lavalamp commented on pkg/api/v1/conversion_test.go:
```diff
@@ -101,3 +102,54 @@ func TestListOptionsConversion(t *testing.T) {
+				api.ResourceCPU: resource.MustParse("4m"),
+			},
+		},
+		{ // Large CPU value should be rounded up to cores.
```
Why?

timstclair replied:
Attempting to use millicore precision at a magnitude that would overflow an int64 I think is just asking for trouble, and I don't see the value in it. I didn't bother to document this since I can't imagine anyone ever running into this (good luck scheduling a pod asking for 10-petacores), but I'm happy to add it for completeness' sake.

lavalamp replied:
Yeah, let's document it. I agree such a pod is unlikely to schedule.
Petacores only happen when you're aggregating pods or nodes.

On Wed, Dec 2, 2015 at 3:05 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/api/v1/conversion_test.go
> <https://github.com/kubernetes/kubernetes/pull/18109#discussion_r46490255>
> :
>
> > +			},
> > +			expected: api.ResourceList{
> > +				api.ResourceMemory:  resource.MustParse("30M"),
> > +				api.ResourceCPU:     resource.MustParse("100m"),
> > +				api.ResourceStorage: resource.MustParse("1G"),
> > +			},
> > +		},
> > +		{ // Nano-core CPU value should be rounded up to millicores.
> > +			input: versioned.ResourceList{
> > +				versioned.ResourceCPU: resource.MustParse("3000023n"),
> > +			},
> > +			expected: api.ResourceList{
> > +				api.ResourceCPU: resource.MustParse("4m"),
> > +			},
> > +		},
> > +		{ // Large CPU value should be rounded up to cores.
>
> Attempting to use millicore precision at a magnitude that would overflow
> an int64 I think is just asking for trouble, and I don't see the value in
> it. I didn't bother to document this since I can't imagine anyone ever
> running into this (good luck scheduling a pod asking for 10-petacores), but
> I'm happy to add it for completeness' sake.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18109/files#r46490255>.
>

timstclair replied:
Done. We don't have any aggregated (or cumulative) quantities in the API, right? I agree this could conceivably be hit in the metrics API.

erictune replied:
I think that CPU should always have millicore precision, no more and no less.  If you wan to make a new api.ResourceFoo and have different precision, then go ahead and do that.

timstclair replied:
I'm happy to remove the rounding for large values, but are you suggesting we reject resources with sub-milli precision? FWIW, the current system behavior is to round up to milli precision.

lavalamp commented:
Looks good-- just one question

lavalamp commented:
LGTM -- squash & fix tests :)

timstclair commented:
FYI, tests are just failing because the dependent PR has not yet been merged.

erictune commented:
I don't think we should round up cpu to less than millicore precision.

timstclair commented:
Done - don't lose millicore precision for large values.

erictune commented:
If we were doing this again, I would preferred to make it so that the apiserver rejects any request that has less than millicore precision.  

Three reasons:

- It seems to go against our API principles to modify a value a user has already set.  It will be confusing in three-way diffs, I think.
- I would rather users know they are doing something that is not allowed, rather than silently get rounded up.
- I want to encourage users to use integers to store the number of millicores, because floating point numbers cause subtle bugs.  By rejecting all cpu requests that cannot be held in an int64 with scale -3, it makes it so that clients know they can safely use an integer representation.

So, I am wondering how bad it would be to change things now to reject values that are not exactly representable in millicores.

@bgrant0607 what do you think about making such a breaking API change?

timstclair commented:
Rebased & broke dependency on https://github.com/kubernetes/kubernetes/pull/17548, so this can be merged first.

@erictune I agree that flat out rejecting sub-milli values seems like the right behaviour, but I'll wait for a decision wrt. the breaking change.

timstclair commented:
Update: I did some more research, and concluded that non-integer memory values are currently allowed, so the changes to validate that Storage & Memory resources are integer values should also be considered a breaking change. Whatever we decide to do with sub-milli CPU values, we should do the same to storage & memory with integer values.

timstclair commented:
I reverted the breaking changes (rejecting non-integer memory & storage), and updated the PR to round ALL resource.Quantities in the API at conversion time. This is effectively a no-op right now, since the resource.Quantity constructor [rounds to millis](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/resource/quantity.go#L199), but this will unblock https://github.com/kubernetes/kubernetes/pull/17548.

I also filed https://github.com/kubernetes/kubernetes/issues/18538 for rejecting invalid values.

PTAL

erictune commented:
lgtm

timstclair commented:
/cc @davidopp

timstclair commented:
Squashed & rebased.

timstclair commented:
Reran `hack/update-all.sh`

tallclair committed:
Enforce minimum resource granularity

Enforce minimum resource granularity of milli-{core, bytes} for Storage,
Memory, and CPU resource types. For Storage and Memory, milli-bytes are
allowed for backwards compatability, but the behavior is
undefinied (depends on docker implementation).

timstclair commented:
Fixed test by adding the following to the conversion function:
```
	if *in == nil {
		return nil
	}
```

timstclair commented:
Applying @erictune 's LGTM since this was supposed to go in before https://github.com/kubernetes/kubernetes/pull/17548
